src/core/utilities/mpi_utils.h

```diff
@@ -23,36 +23,58 @@ std::vector<Data> collect(Data const& data, int mpi_size = 0);
 
 size_t max(size_t local, int mpi_size = 0);
 
+int size();
+
+
 template<typename Data>
 void _collect(Data const* const data, std::vector<Data>& values, size_t send = 1,
               size_t receive = 1)
 {
+    auto gather = [&](auto mpi_type) {
+        MPI_Allgather(data, send, mpi_type, values.data(), receive, mpi_type, MPI_COMM_WORLD);
+    };
+
     if constexpr (std::is_same_v<double, Data>)
-        MPI_Allgather(data, send, MPI_DOUBLE, values.data(), receive, MPI_DOUBLE, MPI_COMM_WORLD);
+        gather(MPI_DOUBLE);
     else if constexpr (std::is_same_v<float, Data>)
-        MPI_Allgather(data, send, MPI_FLOAT, values.data(), receive, MPI_FLOAT, MPI_COMM_WORLD);
+        gather(MPI_FLOAT);
+    else if constexpr (std::is_same_v<int, Data>)
+        gather(MPI_INT);
+    else if constexpr (std::is_same_v<uint32_t, Data>)
+        gather(MPI_UNSIGNED);
+    else if constexpr (std::is_same_v<uint8_t, Data>)
+        gather(MPI_UNSIGNED_SHORT);
     else if constexpr (std::is_same_v<size_t, Data>)
-        MPI_Allgather(data, send, MPI_UINT64_T, values.data(), receive, MPI_UINT64_T,
-                      MPI_COMM_WORLD);
+        gather(MPI_UINT64_T);
     else if constexpr (std::is_same_v<char, Data>)
-        MPI_Allgather(data, send, MPI_CHAR, values.data(), receive, MPI_CHAR, MPI_COMM_WORLD);
+        gather(MPI_CHAR);
     else
-        static_assert("Unhandled MPI data type collection");
+        throw std::runtime_error("Unhandled MPI data type collection");
 }
 
 std::vector<std::string> collectStrings(std::string str, int mpi_size = 0,
                                         std::string null_str = "null");
 
 
 template<typename Data>
-std::vector<std::vector<Data>> collectVector(std::vector<Data> const& data, int mpi_size)
-{
+std::vector<std::vector<Data>> collectVector(std::vector<Data> const& data, int mpi_size,
+                                             bool empty = false)
+{ // can't send empty vectors
+    if (data.size() == 0)
```

Comment:
I think you may want ```MPI_Allgatherv``` instead

https://docs.microsoft.com/en-us/message-passing-interface/mpi-allgatherv-function

Reply:
done